src/data_source/storage/sql.rs

```diff
@@ -2211,4 +2821,314 @@ mod test {
+                    &test_tree,
+                    Snapshot::Index(block_height as u64),
+                    i,
+                )
+                .await
+                .unwrap();
+
+            let (_, proof) = test_tree.lookup(i).expect_ok().unwrap();
+
+            tracing::info!("merkle path {:?}", merkle_path);
+
+            // merkle path from the storage should match the path from test tree
+            assert_eq!(merkle_path, proof.proof.clone(), "merkle paths mismatch");
+        }
+
+        // Lets insert some index again with different block height e.g 0
+        // Our database should then have 2 versions of this leaf node
+        // Update the node so that proof is also updated
+        test_tree.update(0, 99).unwrap();
+        // Also update the merkle commitment in the header
+
+        // data field of the header
+        let test_data = serde_json::json!({ test_tree.header_state_commitment_field() : serde_json::to_string(&test_tree.commitment()).unwrap()});
+        storage
+            .query_opt(
+                "INSERT INTO HEADER VALUES ($1, $2, 't', 0, $3) ON CONFLICT(height) DO UPDATE set data = excluded.data",
+                [
+                    sql_param(&(2_i64)),
+                    sql_param(&"randomstring"),
+                    sql_param(&test_data),
+                ],
+            )
+            .await
+            .unwrap();
+        let (_, proof_bh_2) = test_tree.lookup(0).expect_ok().unwrap();
+        // traversal path for the index.
+        let traversal_path =
+            <usize as ToTraversalPath<U8>>::to_traversal_path(&0, test_tree.height());
+        // Update storage to insert a new version of this code
+
+        <SqlStorage as UpdateStateData<_, MockMerkleTree>>::insert_merkle_nodes(
+            &mut storage,
+            &test_tree,
+            proof_bh_2.proof.clone(),
+            traversal_path.clone(),
+            2,
+        )
+        .await
+        .expect("failed to insert nodes");
+
+        // Find all the nodes of Index 0 in table
+        let ltree_path = LTree::from(traversal_path.iter());
+        let rows = storage
+            .query(
+                "SELECT * from test_tree where pos = $1 ORDER BY created",
+                [sql_param(&ltree_path)],
+            )
+            .await
+            .unwrap();
+
+        let nodes: Vec<_> = rows
+            .map(|res| Node::try_from(res.unwrap()))
+            .try_collect()
+            .await
+            .unwrap();
+        // There should be only 2 versions of this node
+        assert!(nodes.len() == 2, "incorrect number of nodes");
+        assert_eq!(nodes[0].created, 1, "wrong block height");
+        assert_eq!(nodes[1].created, 2, "wrong block height");
+
+        // Now we can have two snapshots for Index 0
+        // One with created = 1 and other with 2
+        // Query snapshot with created = 2
+
+        let path_2 = <SqlStorage as MerklizedStateDataSource<_, MockMerkleTree>>::get_path(
```

Comment:
Need to `get_path` with `Index(1)` to check that the query for the old version still works